### ISSUE TITLE
Do we verify a trust path for artifacts?

### ISSUE THREAD

**Issue by akarasulu** [subutai-io/subos#529]
We just added the signature checksum verification to template artifacts stored in Gorjun. However are we checking that a valid trust chain exists? Meaning if we are just validating the signature is correct and approving then this is not the correct intended operation. We have to make sure the person using the template artifact trusts the owner of the template artifact and there's got to be some kind of authorization given unless of course the template comes from us defaulting to a trusted artifact.

**Comment by tzununbekov** [comment]
@akarasulu No, we just checking signature validity, artifact's hash and matching author. To add web of trust functionality we need a source where we can query user's trust relations with other users. I think Hub as our central authorization server should be responsible for trust chain serving. Need to clarify if Hub have this feature now

**Comment by asanaliev** [comment]
There was some basic implementation for web of trust in Hub (long time ago), along with the custom protocol for a key server. But we haven't utilized this web of trust anywhere since. 
Briefly, there is some basement in Hub ready but we have to review, and may be refactor it to fit our current needs.

**Comment by Dilshat** [comment]
Yes Hub should expose REST for Gorjun to check such trust relations.
Once user managers her artifacts on Gorjun via Hub, Hub can automatically (or prompt user to) establish trust relations between artifacts like owning relation, sharing relation etc.
Gorjun can poll dedicated Hub REST endpoints during trust/signature checks

**Comment by asanaliev** [comment]
I suspect there are similar features in other products and apps. We may research and borrow their working models.

**Comment by akarasulu** [comment]
Yes we will need REST services to look up these authorization documents and to be able to infer trust through chains. For now we know this:

1. The Hub and OD resources are always trusted by the public. 
2. We can also have resources registered with verified organization automatically trusted by the Hub which establishes trust by users Hub->Org->User trust chain 
3. Anything else requires an explicit direct trust by the User->Artifact Owner that gives permissions to use the template or other artifact in their environments.

**Comment by asanaliev** [comment]
Related: https://confluence.critical-factor.com/display/IC/Custom+Protocol

**Comment by asanaliev** [comment]
Seems we might borrow a model used in SSL and certificate chains:
* https://en.wikipedia.org/wiki/X.509
* https://en.wikipedia.org/wiki/Root_certificate
* https://en.wikipedia.org/wiki/Intermediate_Certificate_Authority

Similarities: 
* Our trustdoc (JSON structure) = X509 certificate
* Root CA = Hub
* Intermediate CAs = organizations

https://en.wikipedia.org/wiki/Chain_of_trust: 
> This process results in a chain of trust. The final software can be trusted to have certain properties, because if it had been illegally modified its signature would be invalid, and the previous software would not have executed it. The previous software can be trusted, because it, in turn, would not have been loaded if its signature would have been invalid. The trustworthiness of each layer is guaranteed by the one before, back to the trust anchor - the hardware.

**Comment by asanaliev** [comment]
Also, if I remember, our trustdoc contains min fields only (artifact, signer, etc). We have to review that too and add some more fields. For example, trust expire date.

**Comment by asanaliev** [comment]
Related https://keybase.io/docs/server_security/following: 
> The PGP web of trust has existed for over 20 years. However it is very difficult to use, it requires in-person verifications, and it's hard to know what trust level to assign transitively.
